File: Diagrammatize.Design.Pattern/code/python/Interpreter/Sample2/RepeatCommandNode.py

from Node import Node

# <repeat command> ::= repeat <number> <command list>
class RepeatCommandNode(Node):
    def __init__(self):
        super().__init__()
        self.number = 0
        self.commandListNode = None

    def parse(self, context):
        from CommandListNode import CommandListNode
        context.skipToken("repeat")
        self.number = context.currentNumber()
        context.nextToken()
        self.commandListNode = CommandListNode()
        self.commandListNode.parse(context)

    def __str__(self):
        return "[repeat {} {}]".format(self.number, self.commandListNode)

File: Diagrammatize.Design.Pattern/code/cxx/Interpreter/A1/language/RepeatCommandNode.cpp
#include "RepeatCommandNode.hpp"
#include <sstream>

#include "CommandListNode.hpp"

RepeatCommandNode::~RepeatCommandNode() {}

void RepeatCommandNode::parse(Context& context) {
    context.skipToken("repeat");
    number_ = context.currentNumber();
    context.nextToken();
    commandListNode_.reset(new CommandListNode());
    commandListNode_->parse(context);
}

std::string RepeatCommandNode::toString() {
    std::stringstream ss;
    ss << "[repeat " << number_ << " " 
        << commandListNode_->toString() << "]";
    return ss.str();
}
